### DISCUSSION TITLE
월드상품과 데이터 연동하는법

### DISCUSSION THREAD

**Discussion by JeongYire** [naverz/zepeto-studio-kor#269] [5. Q&A]
월드상품을 연동하고 , IWP버튼을 이용해서 아이템을 구매하는것까진 성공했습니다, 그런데 이걸 다음에 월드를 불러올때 적용시키려면 어떻게 해야하나요? 비소모성 아이템으로 등록해서인지 다시 구매하는게 불가능한걸보니 데이터를 갖고있긴 한 모양인데...  특정 유저가 어떤 월드상품을 갖고있는지 조회할 수 있는 방법이 있을까요?

**Comment by x4cf** [comment]
구매가 완료되면 서버 데이터 스토리지에 저장해서 관리하는게 좋을것 같아요 :)
https://studio.zepeto.me/guides/multiplay-game-datastorage

**Reply by JeongYire** [reply]
// 서버 index 

 async onPurchased(client: SandboxPlayer, receipt: IReceiptMessage) {
        // find purchased user by session id
        const player = this.state.players.get(client.sessionId);        
  
        let itemZ= player.itemZ;
        itemZData .push(receipt.itemId);

        let storage = client.loadDataStorage();
        await storage.set("itemZ",JSON.stringify(itemZ));
    }

 async onJoin(client: SandboxPlayer) {

       [ 중간 생략 ] 

        let itemZData = await storage.get("itemZ") as string;

        if( itemZData == undefined || itemZData == null){
            player.itemZ= new Array<string>();
        }else{
            player.itemZ= JSON.parse(itemZData );
        }

        // client 객체의 고유 키값인 sessionId 를 사용해서 Player 객체를 관리.
        // set 으로 추가된 player 객체에 대한 정보를 클라이언트에서는 players 객체에 add_OnAdd 이벤트를 추가하여 확인 할 수 있음.
        this.state.players.set(client.sessionId, player);
    }


// ClientStarter 


ZepetoPlayers.instance.OnAddedPlayer.AddListener((sessionId: string) => {
                const isLocal = this.room.SessionId === sessionId;
                
                if (!isLocal) {
                    const player: Player = this.currentPlayers.get(sessionId);

                    const myPlayer = ZepetoPlayers.instance.GetPlayer(sessionId);
                    let dataContainer : UnityEngine.GameObject = 
                    UnityEngine.GameObject.Instantiate(this.userDataContainer) as UnityEngine.GameObject;
                    dataContainer.transform.parent = myPlayer.character.transform;
                    dataContainer.GetComponent<WorldPlayer>().player = myPlayer;
                    dataContainer.GetComponent<WorldPlayer>().itemZ= player.itemZ;
                    dataContainer.gameObject.SetActive(true);

                    // [RoomState] player 인스턴스의 state가 갱신될 때마다 호출됩니다.
                    player.OnChange += (changeValues) => this.OnUpdatePlayer(sessionId, player);
                }
            });


이렇게 해서 코드짰을시 소모성아이템을 몇개를 사든 WorldPlayer의 itemZ 의 변수가 변함이 없습니다... 
서버단에서는 저장이 되는걸 확인하였는데 왜 여기선 불러와지지 않을까요?

**Reply by x4cf** [reply]
<img width="633" alt="스크린샷 2022-01-26 오후 2 37 10" src="https://user-images.githubusercontent.com/90679926/151109508-7acb5d17-9549-4a72-b312-6d12e74ded56.png">

**Reply by JeongYire** [reply]
일단 고치긴 하였습니다만, 이 문제는 아닌거같습니다. 저 상태로도 myPlayer 변수는 제대로 불러와졌기때문입니다

**Reply by x4cf** [reply]
dataContainer가 잘 불러와진다면 코드만 보고는 무엇이 문제인지 잘 모르겠습니다.
서버에 저장하고 불러오는 부분이 문제가 있지 않을까 싶습니다.

**Reply by JeongYire** [reply]
#112 
이 문제가 아직 해결이 안되어서 그런거 인거같습니다. 아쉽지만 이 이슈가 해결되고 나서 시도해봐야겠네요

**Reply by x4cf** [reply]
문제가 되는 부분도 있나보군요!
저는 서버 데이터 스토리지가 잘 작동되고있는데 운좋게 문제되는부분은 모두 피했었나봅니다.

**Reply by JeongYire** [reply]
해결됐습니다... 로컬 플레이어 생성시에 Player 스키마를 참조하지 않아 발생한 문제였습니다 
#112 문제는 로컬환경이아니라 앱내에서 발생하는 문제같네요

**Reply by x4cf** [reply]
해결되었다니 다행입니다 :D
